src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -62,15 +67,341 @@ int
+}
+
+// Utility function for device class to verify that the C xclbin handle is valid
+// Needed when the C API for device tries to load an xclbin using C pointer to xclbin
+namespace xrt {
+namespace Xclbin {
+bool
+is_valid(xrtXclbinHandle handle)
+{
+  bool rtn = (xclbins.find(handle) != xclbins.end());
+  return rtn;
+}
+}
+};
+
+xrtXclbinHandle
+xrtXclbinAllocFilename(const char *filename)
+{
+  try {
+    auto xclbin = std::make_shared<xrt::xclbin_impl>(filename);
+    auto handle = xclbin.get();
+    xclbins.emplace(std::make_pair(handle, std::move(xclbin)));
+    return handle;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    errno = ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    errno = 0;
+  }
+  return nullptr;
+}
+
+xrtXclbinHandle
+xrtXclbinAllocRawData(const void *data, const int size)
+{
+  try {
+    char *data_c = (char *) data;
+    std::vector<char> raw_data(data_c, data_c + size);
+    auto xclbin = std::make_shared<xrt::xclbin_impl>(raw_data);
+    auto handle = xclbin.get();
+    xclbins.emplace(std::make_pair(handle, std::move(xclbin)));
+    return handle;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    errno = ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    errno = 0;
+  }
+  return nullptr;
+}
+
+int
+xrtXclbinFreeHandle(xrtXclbinHandle handle)
+{
+  try {
+    free_xclbin(handle);
+    handle = nullptr;
+    return 0;
+  }
+  catch (const xrt_core::error &ex) {
     xrt_core::send_exception_message(ex.what());
+    return ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
     return -1;
   }
 }
+
+int
+xrtXclbinGetCUNames(xrtXclbinHandle handle, char **names, int *numNames)
+{
+  try {
+    auto xclbin = get_xclbin(handle);
+    const std::vector<std::string> cuNames = xclbin->get_cu_names();
+    // populate numNames if memory is allocated
+    if (numNames != nullptr) *numNames = cuNames.size();
+    // populate names if memory is allocated
+    if (names != nullptr) {
+      auto index = 0;
+      for (auto &&name: cuNames) {
+        std::strcpy(names[index++], name.c_str());
+      }
+    }
+    return 0;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    return (errno = ex.get());
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    return (errno = 0);
+  }
+}
+
+int
+xrtXclbinGetXSAName(xrtXclbinHandle handle, char *name, int *size)
+{
+  try {
+    auto xclbin = get_xclbin(handle);
+    const std::string &xsaName = xclbin->get_xsa_name();
+    // populate size if memory is allocated
+    if (size != nullptr)
+      *size = xsaName.size();
+    // populate name if memory is allocated
+    if (name != nullptr)
+      std::strcpy(name, xsaName.c_str());
```

Comment:
The signature should be
xrtXclbinGetXSAName(handle, char* name, size_t size, size_t* ret_size);

If ret_size is not nullptr, then populate with number of characters needed for name.  size is the size of the name string user allocate, so never copy more than size.

strncpy, user should provide the length of the name string.  At least here the user has a chance to allocate his name string per returned ret_size for use case 1 where size==0 or name==nullptr, but he may allocate less than needed and that's ok, or you can make it an error if you think it is not okay.